tests/vmemcache_fragmentation.c

```diff
@@ -0,0 +1,280 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * vmemcache_fragmentation.c -- fragmentation test source
+ */
+
+#include "vmemcache_tests.h"
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <libvmemcache.h>
+#include <limits.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#define ALLOWED_RATIO 0.95
+#define MAX_KEYSIZE 30
+
+typedef struct {
+	size_t allocated_size;
+	size_t index;
+	bool evicted;
+
+} on_evict_info;
+
+typedef struct {
+	size_t pool_size;
+	size_t segment_size;
+	size_t val_max;
+	char dir[PATH_MAX];
+	int seconds, minutes, hours;
+} test_params;
+
+/* Generate random size value */
+static size_t
+get_rand_size(size_t val_max, size_t segment_size)
+{
+	size_t val_size =
+	    (1 + (size_t) rand() / (RAND_MAX / (val_max / segment_size) + 1)) *
+	    segment_size;
+
+	assert(val_size <= val_max);
+	assert(val_size >= segment_size);
+	assert(val_size % segment_size == 0 &&
+		"put value size must be a multiple of segment size");
+
+	return val_size;
+}
+
+/* Calculate allocated space size on evict */
+static void
+on_evict(VMEMcache * cache, const char *key, size_t key_size, void *arg)
+{
+	on_evict_info *info = (on_evict_info *) arg;
+	info->evicted = true;
+	char buf[1];
+	size_t vsize;
+
+	ssize_t ret;
+	if ((ret =
+		vmemcache_get(cache, key, key_size, &buf, 1, 0, &vsize)) != 1) {
+		FATAL("vmemcache_get: invalid ret (actual %zu, expected 1). %s",
+		ret, strerror(errno));
+	}
+	info->allocated_size -= vsize;
+}
+
+/* Return 0 if timeout occurs, 1 otherwise */
+static int
+timeout(time_t start, int seconds, int minutes, int hours)
+{
+	time_t end;
+	int diff;
+	int timeout = seconds + 60 * minutes + 3600 * hours;
+
+	/* run indefinitely */
+	if (timeout == 0)
+		return 1;
+
+	time(&end);
+
+	diff = (int)difftime(end, start);
+	if (diff >= timeout) {
+		return 0;
+	}
+	return 1;
+}
+
+/* Parse cmd args */
+static test_params
+parse_args(int argc, char **argv)
+{
+	test_params p = {.pool_size = VMEMCACHE_MIN_POOL,
+		.segment_size = 16,
+		.val_max = 0,
+		.dir = "/tmp",
+		.seconds = 0,
+		.minutes = 0,
+		.hours = 0
+	};
+	size_t val_max_factor = 70;
+
+	static const char *usage_str = "usage: %s "
+		"[-h] "
+		"[-p <pool_size>] "
+		"[-s <segment_size>] "
+		"[-v <val_max_factor>] "
+		"[-t <timeout_seconds>] "
+		"[-m <timeout_minutes>] "
+		"[-o <timeout_hours>] "
+		"[-d <dir>]\n ";
+
+	const char *optstr = "hp:s:v:t:m:o:d:";
+	int opt;
+	while ((opt = getopt(argc, argv, optstr)) != -1) {
+		char *endptr;
+		switch (opt) {
+		case 'h':
+			printf(usage_str, argv[0]);
+			exit(0);
+		case 'p':
+			p.pool_size = (size_t)(strtoumax(optarg, &endptr, 10));
+			if (strlen(endptr) != 0)
```

Comment:
Missing error handling. And you don't even need to call strlen for the case you want to check - "*endptr != 0" is enough.
Since this code is copy pasted multiple times I'd recommend adding a helper function.